launch_frontend/test/launch_frontend/test_substitutions.py

```diff
@@ -0,0 +1,112 @@
+
+
+def test_text_only():
+    subst = default_parse_substitution('yes')
+    assert len(subst) == 1
+    assert subst[0].perform(None) == 'yes'
+    subst = default_parse_substitution('10')
+    assert len(subst) == 1
+    assert subst[0].perform(None) == '10'
+    subst = default_parse_substitution('10e4')
+    assert len(subst) == 1
+    assert subst[0].perform(None) == '10e4'
+    subst = default_parse_substitution('10e4')
+    assert len(subst) == 1
+    assert subst[0].perform(None) == '10e4'
+
+
+@expose_substitution('test')
+def parse_test_substitution(data):
+    if not data or len(data) > 1:
+        raise RuntimeError()
+    return TextSubstitution(text=''.join([i.perform(None) for i in data[0]]))
+
+
+def test_text_with_embedded_substitutions():
+    subst = default_parse_substitution('why_$(test asd)_asdasd_$(test bsd)')
+    assert len(subst) == 4
+    assert subst[0].perform(None) == 'why_'
+    assert subst[1].perform(None) == 'asd'
+    assert subst[2].perform(None) == '_asdasd_'
+    assert subst[3].perform(None) == 'bsd'
+
+# TODO(ivanpauno): Don't deppend on substitution parsing methods for testing the interpolator.
+# Write some dummy substitutions and parsing methods instead.
+
+
+def test_substitution_with_multiple_arguments():
+    subst = default_parse_substitution('$(env what heck)')
+    assert len(subst) == 1
+    subst = subst[0]
+    assert subst.name[0].perform(None) == 'what'
+    assert subst.default_value[0].perform(None) == 'heck'
+
+
+def test_escaped_characters():
+    subst = default_parse_substitution(r'$(env what/\$\(test asd\\\)) 10 10)')
+    assert len(subst) == 2
+    assert subst[0].name[0].perform(None) == 'what/$(test'
+    assert subst[0].default_value[0].perform(None) == r'asd\)'
+    assert subst[1].perform(None) == ' 10 10)'
+
+
+def test_nested_substitutions():
+    subst = default_parse_substitution('$(env what/$(test asd) 10) 10 10)')
+    assert len(subst) == 2
+    assert len(subst[0].name) == 2
+    assert subst[0].name[0].perform(None) == 'what/'
+    assert subst[0].name[1].perform(None) == 'asd'
+    assert subst[0].default_value[0].perform(None) == '10'
+    assert subst[1].perform(None) == ' 10 10)'
+
+
+def test_quoted_nested_substitution():
+    subst = default_parse_substitution(
+        'go_to_$(env WHERE asd)_of_$(env '
+        "'something $(test 10)')"
+    )
+    assert len(subst) == 4
+    assert subst[0].perform(None) == 'go_to_'
+    assert subst[1].name[0].perform(None) == 'WHERE'
+    assert subst[1].default_value[0].perform(None) == 'asd'
+    assert subst[2].perform(None) == '_of_'
+    assert subst[3].name[0].perform(None) == 'something '
+    assert subst[3].name[1].perform(None) == '10'
+    assert subst[3].default_value[0].perform(None) == ''
+
+
+def test_double_quoted_nested_substitution():
+    subst = default_parse_substitution(
+        r'$(env "asd_bsd_qsd_$(test "asd_bds")" "$(env DEFAULT)_qsd")'
+    )
+    assert len(subst) == 1
```

Comment:
@ivanpauno missing test?

Reply:
Thanks! It was actually a problematic case